#Genome Wide Association pipeline for Potato#

*Calling libraries*

````{r}
   library(devtools)
   library(BGData)
   library(BGLR)
````
*Setting up the parameters*

````{r}
  rm(list=ls())
  setwd("/Users/fenciso/Desktop/Thesis/GWAS_GS/r")
  dataFile<-'LB_Z_GEN_PHEN.RData'
  GMatrixFile='G_TETRA.RData'
  mapFile='genos.RData'
  traitName<-"LB"
  
## QC filter parameters
  scaleCol<-TRUE
  centerCol<-TRUE
  minMAF<-.03
  maxFreqNA.MRK<-.2
  maxFreqNA.Genos<-0.15
````

*Loading the files*

````{r}
  load(dataFile)
  load(GMatrixFile)
  load(mapFile)
  genoFile<-Z_filter2

## QC and subsetting of columns from X and rows from MAP
 


# Basic QC, removing markers based on MAF and % of NAs
 
tmp<-which(MAP$maf<minMAF | MAP$freqNA>maxFreqNA.MRK)
 if(length(tmp)>0){
    genoFile<-genoFile[,-tmp]
 }

# Removing genotypes with too many NAs
 tmp<-which(rowMeans(is.na(genoFile))>maxFreqNA.Genos)
 if(length(tmp)>0){ genoFile<-genoFile[-tmp,] }

 # Native imputation
 for(i in 1:ncol(genoFile)){
    tmp<-mean(genoFile[,i],na.rm=T)
    genoFile[,i]<-ifelse(is.na(genoFile[,i]),tmp,genoFile[,i])
 }

X2<-genoFile

## checking if the X, G and MAP files match

stopifnot(all.equal(names(blups),rownames(X2)))

stopifnot(all(names(blups)%in%rownames(G)))
tmp=!rownames(G)%in%names(blups)
if(any(tmp)){G=G[!tmp,!tmp]}
stopifnot(all(names(blups)%in%rownames(G)))
stopifnot(all(rownames(G)%in%names(blups)))
tmp=order(as.integer(factor(x=rownames(G),ordered=TRUE,levels=names(blups))))
G=G[tmp,tmp]

stopifnot(all.equal(names(blups),rownames(G)))
stopifnot(all.equal(names(blups),colnames(G)))

 y=scale(blups)
 PC=eigen(G)$vectors[,1:2]
 #DATA=new('BGData',geno=X,pheno=data.frame(y,PC))



 GWAS=matrix(nrow=ncol(X2),ncol=4,NA)

index = match(colnames(X2),MAP$Name)
MAP2 = MAP[index,]

 for(i in 1:ncol(X2)){
    fm<-lm(y~PC+X2[,i])
    GWAS[i,]<-summary(fm)$coef[4,]
    print(i)
 }



#z4=apply(X2,2, function(x) (as.numeric(x>3)))
  
## Without PC

GWAS1=matrix(nrow=ncol(X2),ncol=4,NA)
dimnames=c(rownames=colnames(X2))
GWAS2=matrix(nrow=ncol(X2),ncol=4,NA)
GWAS3=matrix(nrow=ncol(X2),ncol=4,NA)
GWAS4=matrix(nrow=ncol(X2),ncol=4,NA)


for (i in 1:ncol(X2)){
  z1=as.numeric(X2[,i]>0)
  if ((sum(z1)/length(z1))!=1){
    fm<-lm(y~z1) 
    GWAS1[i,]<-summary(fm)$coef[2,]
    print(i)
   
    z2=as.numeric(X2[,i]>1)
    if ((sum(z2)/length(z2))!=0 & (sum(z2)/length(z2))!=1){
        fm1<-lm(y~z2+z1)
        GWAS2[i,]<-summary(fm1)$coef[2,]
        print(i)
        
        z3=as.numeric(X2[,i]>2)
        if ((sum(z3)/length(z3))!=0 & (sum(z3)/length(z3))!=1){
          fm2<-lm(y~z3+z2+z1)
          GWAS3[i,]<-summary(fm2)$coef[2,]
          
          z4=as.numeric(X2[,i]>3)
          if ((sum(z4)/length(z4))!=0 & (sum(z4)/length(z4))!=1){
            fm3<-lm(y~z4+z3+z2+z1)
            GWAS4[i,]<-summary(fm3)$coef[2,]
            
            }
          }
        }
      }
    }


  
  

  z2=as.numeric(X2[,6]>1)
  fm1<-lm(y~z1+z2)
  GWAS2[1,]<-summary(fm1)$coef[3,]
  print(i)
}
  
  z3=as.numeric(X2[,i]>2)
  fm2<-lm(y~z3+z2+z1)
  GWAS3[i,]<-summary(fm2)$coef[,4]
  }
  
  for (i in 1:ncol(X2)){
  z4=as.numeric(X2[,i]>3)
  fm3<-lm(y~z4+z3+z2+z1)
  GWAS4[i,]<-summary(fm3)$coef[4,]

}


 
plot(main="GWAS for Late Scab", numeric()~numeric(),xaxt="n",xlab="Chromosome",xlim=c(0,nrow(GWAS)),ylim=c(0,max(max(-log(GWAS[,4],base=10),na.rm=T),-log10(.01/nrow(GWAS)))), ylab="-Log10(p_value)")

#plot(main="GWAS for Scab", numeric()~numeric(),xaxt="n",xlab="Chromosome",xlim=c(0,nrow(GWAS)),ylim=c(0,max(-log(GWAS[,4]),na.rm=T)), ylab="-Log10(p_value)")

chrlist=names(table(MAP2$PM_chr_v4_03))
lnol=0
posi = c()
for (i in 1:length(chrlist)){
  ini=lnol+1
  temp=which(MAP2$PM_chr_v4_03==chrlist[i])
  posi[i]=length(temp)
  lnol=ini+length(temp)-1
  y=-log(GWAS[temp,4])
  x=ini:lnol
  points(x=x,y=y,col=i, cex=0.5)
  print(c(ini,lnol))
  
}
posi2 = cumsum(posi)
tt = c(0,posi2[-length(posi2)])+posi/2
#axis(1,tt,chrlist,cex=0.5,las=2)
axis(1,tt,c(1:13),cex=0.2, las=2)
bt= -log10(0.01/nrow(GWAS))
abline (h=bt,col="red")
````
[Back](https://github.com/fenciso13/Potato_MSU/)
