
#Genome Wide Association analysis

*Calling libraries*

````{r}
   library(devtools)
   library(BGData)
   library(BGLR)
````
*Setting up parameters*

````{r}
rm(list=ls())
setwd("/Users/fenciso/Desktop/Thesis/GWAS_GS/r")
dataFile<-'LB_X_GEN_PHEN.RData'
GMatrixFile='G_TETRA.RData'
mapFile='genos.RData'
traitName<-"LB"
````

*Loading the files*

````{r}
load(dataFile)
load(GMatrixFile)
load(mapFile)
genoFile=X_filter2
````

*Checking if the X, G and MAP files match*

````{r}
stopifnot(all.equal(names(blups),rownames(genoFile)))
stopifnot(all(names(blups)%in%rownames(G)))

tmp=!rownames(G)%in%names(blups)

if(any(tmp)){G=G[!tmp,!tmp]}

stopifnot(all(names(blups)%in%rownames(G)))
stopifnot(all(rownames(G)%in%names(blups)))

tmp=order(as.integer(factor(x=rownames(G),ordered=TRUE,levels=names(blups))))
G=G[tmp,tmp]

stopifnot(all.equal(names(blups),rownames(G)))
stopifnot(all.equal(names(blups),colnames(G)))
````

*Performing the GWA*

````{r}
y=scale(blups) # centering the blups
PC=eigen(G)$vectors[,1:2]
 #DATA=new('BGData',geno=X,pheno=data.frame(y,PC))

GWAS=matrix(nrow=ncol(genoFile),ncol=4,NA)# empty matrix in which the results will be store

for(i in 1:ncol(genoFile)){
   fm<-lm(y~PC+genoFile[,i])
   GWAS[i,]<-summary(fm)$coef[4,]
   print(i)
 }
````

*Generating the Manhattan plot*

````{r}

index=match(colnames(genoFile),MAP$Name)
MAP2=MAP[index,] # filtering the map file according with the lines present in the filtered genotype file

par(mar = rep(2, 4)) # This line avoid the margin problems
plot(main="GWAS for Late blight", numeric()~numeric(),xaxt="n",xlab="Chromosome",xlim=c(0,nrow(GWAS)),ylim=c(0,max(max(-log(GWAS[,4],base=10),na.rm=T),-log10(.01/nrow(GWAS))+8)), ylab="-Log10(p_value)")

chrlist=names(table(MAP2$PM_chr_v4_03))
lnol=0
posi = c()

# Using the next loop we generated the manhattan plot ordered by chromosome
for (i in 1:length(chrlist)){
  ini=lnol+1
  temp=which(MAP2$PM_chr_v4_03==chrlist[i])
  posi[i]=length(temp)
  lnol=ini+length(temp)-1
  y=-log(GWAS[temp,4])
  x=ini:lnol
  points(x=x,y=y,col=i, cex=0.5)
  print(c(ini,lnol))
  }

# Adding the chromosome number
posi2 = cumsum(posi)
tt = c(0,posi2[-length(posi2)])+posi/2
axis(1,tt,c(1:13),cex=0.2, las=2)
bt= -log10(0.01/nrow(GWAS))
abline (h=bt,col="red")
````
![screenShot](https://github.com/fenciso13/GWAS_and_GS/blob/master/pdf/GWAS.png)

````{r}
##############################################################

GWAS=matrix(nrow=ncol(genoFile),ncol=13,NA)
colnames(GWAS)=c("DipA","DipD","TetraA","Z1","Z2","Z3","Z4","DipA-DipD","TetraA-DipA","Z2-Z1","Z3-Z2","Z4-Z3","Z4-TetraA")
dimnames=c(rownames=colnames(genoFile))
rownames(GWAS)=dimnames
fm0=lm(y~1)

for (i in 1:ncol(genoFile)){
  xi=genoFile[,i]
  z1=as.numeric(xi>0)
  z2=as.numeric(xi>1)
  z3=as.numeric(xi>2)
  z4=as.numeric(xi>3)
  DipA=ifelse(xi==0,-1,ifelse(xi>3,1,0))
  DipD=as.integer(DipA==0)
  
  fmDipA=lm(y~DipA)
  fmDipAD=lm(y~DipA+DipD)
  fmTetraA=lm(y~xi)
  fmZ1=lm(y~z1)
  fmZ2=lm(y~z2)
  fmZ3=lm(y~z3)
  fmZ4=lm(y~z4)
  GWAS[i,1]=anova(fmDipA,fm0)[[6]][2]
  GWAS[i,2]=anova(fmDipAD,fm0)[[6]][2]
  GWAS[i,3]=anova(fmTetraA,fm0)[[6]][2]
  GWAS[i,4]=anova(fmZ1,fm0)[[6]][2]
  GWAS[i,5]=anova(fmZ2,fm0)[[6]][2]
  GWAS[i,6]=anova(fmZ3,fm0)[[6]][2]
  GWAS[i,7]=anova(fmZ4,fm0)[[6]][2]
  GWAS[i,8]=anova(fmDipA,fmDipAD)[[6]][2]
  GWAS[i,9]=anova(fmTetraA,fmDipAD)[[6]][2]
  GWAS[i,10]=anova(fmZ2,fmZ1)[[6]][2]
  GWAS[i,11]=anova(fmZ3,fmZ2)[[6]][2]
  GWAS[i,12]=anova(fmZ4,fmZ3)[[6]][2]
  GWAS[i,13]=anova(fmZ4,fmTetraA)[[6]][2]
}

````
[Next_module](https://github.com/fenciso13/GWAS_and_GS/blob/master/modules/6.%20GS.md)

[Back](https://github.com/fenciso13/GWAS_and_GS)
