#Reading and recoding genotypes:#

```{r}
rm(list=ls())
setwd("/Users/fenciso/Desktop/Thesis/GWAS_GS/r")
fileIn<-"/Users/fenciso/Desktop/Thesis/GWAS_GS/r/New_genotype/MSU_GWAS_LB_Scab_189_samples_4859_AABB_filter.txt"
fileGeno<-"/Users/fenciso/Desktop/Thesis/GWAS_GS/r/New_genotype/MSU_GWAS_LB_Scab_189_samples_4859_AABB_recoded.RData"
fileMap<-"/Users/fenciso/Desktop/Thesis/GWAS_GS/r/New_genotype/MSU_GWAS_LB_Scab_189_samples_4859_AABB_map.RData" 
```
*Setting up paremeters*

```{r}
nColSkip<-6 # This vector indicates the number of columns that you want to move to the map file
idCol<-2 # SNP IDs
na.string='.' # This is the character that identifies NAs in the current file
sep.string=''
ploidy<-4 # Use 2 for diploids and 4 for tetraploids
lineNm=c() # add the names of the lines here ie."stub-71",  "stub-640", "stub-10"
```
*Modifying the input file:*

    a. Generating the geno and map file
```{r}
DATA<-read.table(file=fileIn,sep='\t',header=T,stringsAsFactors=F)
colnames(DATA)<-gsub(".", "-", colnames(DATA), fixed = TRUE)

MAP<-DATA[,1:nColSkip]
colnames(MAP)=gsub("-","_",colnames(MAP))

DATA<-as.matrix(DATA[,-(1:nColSkip)])
tmp<-paste0(rep('-',ploidy),collapse='')

DATA<-ifelse(DATA==na.string,tmp,DATA)
na.string<-tmp
```
    b. Adding SNP allele frequency stimates columns in the map file

```{r}
n<-ncol(DATA)
p<-nrow(DATA)
X<-matrix(nrow=n,ncol=p,NA)
rownames(X)<-colnames(DATA)
colnames(X)<-MAP[,idCol]
MRKID<-rep(NA,p)
MAP$freqNA<-NA
MAP$alleleOne<-NA
MAP$alleleTwo<-NA
MAP$freqNA<-NA
MAP$freqAlleleOne<-NA
MAP$maf<-NA
```
    c. Calculating the stimates
```{r}
for(i in 1:p){
    x<-DATA[i,,drop=TRUE]
    tmp<-x==na.string
    names(x)<-NULL
    geno<-matrix(data=unlist(strsplit(x,split='')),ncol=ploidy,byrow=T)
    tmp<-which(x==na.string)
    if(length(tmp)>0){
       geno[tmp,]<-NA
       x[tmp]<-NA
     }
    alleles<-unique(unlist(strsplit(names(table(x)),split='')))
    if(length(alleles)>0){
    if(length(alleles)>2){ stop(paste('Marker',i,'has more than two alleles!')) }
        MAP$alleleOne[i]<-alleles[1]
    if(length(alleles)>1){
        MAP$alleleTwo[i]<-alleles[2]
    }
    }
    MAP$freqNA[i]<-mean(is.na(x))
    tmp<-MAP$alleleOne[i]
    z<-geno[,1]==tmp
    for(j in 2:ploidy){
        z<-z+(geno[,j]==tmp)
    }
    X[,i]<-z
    MAP$freqAlleleOne[i]<-mean(z,na.rm=T)/ploidy
    print(i)
} 
MAP$maf<-ifelse(MAP$freqAlleleOne>0.5,1-MAP$freqAlleleOne,MAP$freqAlleleOne)
```

*Ordering chromoses*

```{r}
MAP$PM_chr_v4_03<-gsub("chr00","chr13",MAP$PM_chr_v4_03)
index=order(MAP$PM_chr_v4_03,MAP$PM_pos_v4_03) # columns from the map file
MAP=MAP[index,]
X=X[,index]

```
*Removing lines*

```{r}
X=X[!rownames(X)%in%lineNm,]
```
*Generating the genotype file considering a diploid additive effect*

```{r}
Z<-X
Z<- ifelse(0<X & X<4, 1,ifelse(X==4,2,X))
```
*Saving the files generated in this module*

```{r}
save(MAP,X,Z,file='/Users/fenciso/Desktop/Thesis/GWAS_GS/r/genos.RData')
```
[Back](https://github.com/fenciso13/Potato_MSU/)
